core/src/main/java/org/apache/iceberg/inmemory/InMemoryCatalog.java

```diff
@@ -138,13 +138,13 @@ public boolean dropTable(TableIdentifier tableIdentifier, boolean purge) {
 
   @Override
   public List<TableIdentifier> listTables(Namespace namespace) {
-    if (!namespaceExists(namespace) && !namespace.isEmpty()) {
+    if (!namespaceExists(namespace)) {
       throw new NoSuchNamespaceException(
           "Cannot list tables for namespace. Namespace does not exist: %s", namespace);
     }
 
     return tables.keySet().stream()
-        .filter(t -> namespace.isEmpty() || t.namespace().equals(namespace))
+        .filter(t -> t.namespace().equals(namespace))
```

Comment:
This looks like the correct behavior, but I think there are problems with the implicit existence of the empty namespace.

`listTables` currently assumes that the empty namespace exists. However, to create a table in that namespace, you'd need to create it. We need to solve that problem, but probably not in this PR. For this PR, `listTables` should return an empty set of tables and we can clean up not needing to create the root/empty namespace later.